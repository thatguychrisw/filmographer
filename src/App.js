import React, { useState, useEffect } from 'react';
import ActorSearch from './actor-search';
import FilmographyList from './filmography-list-view';

function useActorApi () {
    const [movies, setMovies] = useState([]);
    const [isLoading, setLoading] = useState(false);
    const [criteria, setCriteria] = useState(null);

    useEffect(() => {
        (async () => {
            if (criteria === null) return;

            setLoading(true);

            try {
                const response = await fetch('https://jsonplaceholder.typicode.com/albums');
                const filmography = await response.json();

                const random = Math.floor(Math.random() * 15);
                let randomizedMovies = filmography.slice(0, random);
                setMovies(randomizedMovies);

                setLoading(false);
            } catch (e) {
                console.log('An error occurred ', e);
            }

        })();
    }, [criteria]);

    return [{isLoading, movies}, setCriteria];
}

function App () {
    const [{isLoading, movies}, fetchFilmography] = useActorApi();

    return (
      <div className="flex h-screen">
          <div className="m-auto w-11/12">
              <header>
                  <h1 className="text-center text-6xl p-4 text-white">Filmographer</h1>
              </header>

              <ActorSearch onSearch={fetchFilmography}/>
              {isLoading && <div>Loading...</div>}

              <FilmographyList movies={movies}/>
          </div>
      </div>
    );
}

export default App;
