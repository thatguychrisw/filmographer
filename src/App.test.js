import React from 'react';
import { render, fireEvent, waitForElement } from '@testing-library/react'
import App from './App';

const pressEnterOn = (node) => fireEvent.keyDown(node, { key: 'Enter', code: 13 });
const changeOn = (node, value) => fireEvent.change(node, { target: { value } });
const mockOneRequest = (body) => fetch.mockResponseOnce(JSON.stringify(body));

it('renders without crashing', () => {
    const {container} = render(<App />);
    expect(container.firstChild).toMatchSnapshot();
});

it('attempts to fetch the filmography for an actor', async () => {
    const {getByTestId} = render(<App />);

    let filmographyStub = [{title: "Test"}];
    mockOneRequest(filmographyStub);

    const queryInputField = getByTestId('fetch-filmography');
    changeOn(queryInputField, 'test');
    pressEnterOn(queryInputField);

    await waitForElement(() => getByTestId('filmography-list'));

    const list = getByTestId('filmography-list').querySelectorAll('li');
    expect(list.length).toEqual(filmographyStub.length);
});

