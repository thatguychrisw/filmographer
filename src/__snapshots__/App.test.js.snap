// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="flex h-screen"
>
  <div
    class="m-auto w-11/12"
  >
    <header>
      <h1
        class="text-center text-6xl p-8 text-white"
      >
        Filmographer
      </h1>
    </header>
    <div
      class="bg-white rounded w-5/12 mt-4 p-2 flex flex-col m-auto shadow-xl"
    >
      <div
        class="flex"
      >
        <input
          class="w-9/12 outline-none p-2 border-gray-400"
          data-testid="fetch-filmography"
          placeholder="Search for an actor"
          type="text"
        />
      </div>
    </div>
    <div
      class="rounded w-2/3 mt-4 p-2 flex flex-col m-auto"
    />
  </div>
</div>
`;
